src/queryBuilder/QueryBuilder.js

```diff
@@ -0,0 +1,138 @@
+      flat[encode(to)] = node
+    },
+    reparentPaths(node, oldParent, newParent) {
+      treeUtils.walk(node => {
+        Tree.updatePath(node, [
+          ...newParent,
+          ...node.path.slice(oldParent.length),
+        ])
+      })(node)
+    },
+  }))
+  return {
+    getNode: Tree.getNode,
+    add: tree => Tree.add(tree.path, observable(blankNode())),
+    remove: (tree, node) => Tree.remove(node.path),
+    join: (tree, join) => Tree.mutate(tree.path, { join }),
+    mutate: Tree.mutate,
+    typeChange(node, type) {
+      let path = node.path
+      let parentPath = _.dropRight(1, node.path)
+      let tree = Tree.getNode(parentPath)
+      let index = tree.children.indexOf(node)
+      Tree.remove(path)
+      Tree.add(
+        parentPath,
+        observable({
+          key: node.key,
+          type,
+          field: node.field,
+        })
+      )
+      let newNode = Tree.getNode(path)
+      // Move to same index
+      tree.children.remove(newNode) // pop since add does a push
+      tree.children.splice(index, 0, newNode)
+    },
+    move(tree, node, targetTree, index) {
+      // if tree != target tree, do remove/add/move for reactors
+      // need to update paths
+      tree.children.splice(tree.children.indexOf(node), 1)
+      targetTree.children.splice(index, 0, node)
+    },
+    indent(tree, node, skipDefaultNode) {
+      // Reactors:
+      //   OR -> And, nothing
+      //   AND -> OR, others if has value
+      //   to/from NOT, others if has value
+      if (!tree) {
+        // Indent in place, only needs to happen at root and only works if node is a group
+        let key = randomString()
+        let newGroup = observable({
+          key,
+          join: node.join,
+          children: node.children,
+          path: [...node.path, key],
+        })
+        node.children = [newGroup]
+        node.join = oppositeJoin(node.join)
+
+        Tree.reparentPaths(newGroup, node.path, newGroup.path)
+        Tree.add(node.path, observable(blankNode()))
+      } else {
+        let newGroup = observable({
+          key: randomString(),
+          join: oppositeJoin(tree.join),
+          children: [node],
+        })
+        replaceOn(tree.children, node, newGroup)
+
+        // Update paths
+        Tree.updatePath(newGroup, [..._.dropRight(1, node.path), newGroup.key])
+        Tree.reparentPaths(node, _.dropRight(1, node.path), newGroup.path)
+
+        if (!skipDefaultNode) Tree.add(newGroup.path, observable(blankNode()))
+
+        return newGroup
+      }
+    },
+  }
+}
+
+export default DDContext(
+  Component(
+    ({ tree: iTree, types: iTypes }, { types = iTypes, tree = iTree }) => ({
+      types,
+      state: observable({
+        adding: false,
+        ...ContextureClientBridge(types, tree),
+      }),
+    }),
+    ({ state, path, fields, types = {} }) => (
+      <Provider fields={fields} types={types}>
+        <div style={{ background }}>
```

Comment:
For the background, absolutely. This control can't have a transparent background because there are a few places where it uses divs to cover lines (haven't found a better solution 😅), so this needs to be able to be passed in